GetStream/stream-chat-android
# [3727] Compose bi directional pagination

### 🎯 Goal

Add bi-directional pagination to the Compose SDK with scroll to selected message and scroll to bottom.

### 🛠 Implementation details

Updated `ChannelLogic` which determines if the pagination has reached either end or start of the messages list. Also added checks when receiving or sending new messages, so that the new message isn't added to the list if the newest messages are not loaded to avoid breaking pagination of newer messages.

Introduced new pagination handler for reaching the end of new messages to `MessageList` composable. Also added `ScrollToPositionState` inside `MessagesState` which notifies if we are loading the focused messages or the newest messages if the user wants to scroll to the bottom of the list, or if we should scroll to the focused message or bottom of the list. Added scroll to bottom handler to `MessageList`.

### 🎨 UI Changes

<table>
<thead>
<tr>
<th>Before</th>
<th>After</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<video src="https://user-images.githubusercontent.com/38032787/176719148-72f40164-bf9b-433e-9f5e-190f44cdc6ea.mp4" controls="controls" muted="muted" />
</td>
<td>
<video src="https://user-images.githubusercontent.com/38032787/176719135-254b3323-06c0-4064-9f9a-c2bba4cf9157.mp4" controls="controls" muted="muted" />
</td>
</tr>
</tbody>
</table>


### 🧪 Testing

1. Build this branch and enter a channel that has more than one page
2. Scroll n number of pages and reply to a message
4. Exit the channel and enter it again, the message history will be reset
5. Click on the quoted message which should load the message and a page that surrounds the message
6. Scroll up and down to ensure that the bi directional pagination is working
7. Try sending messages from the current user and from a different device, messages shouldn't be added to the bottom of the list, otherwise pagination will break
8. Click on scroll to bottom button, if the newest messages aren't present they should be loaded and the list should scroll to the bottom

Test that the ui-components sdk still has functioning two way pagination since there were changes how the end of list is determined. Scroll to bottom will come in a separate pr.

### ☑️Contributor Checklist

#### General

#### Code & documentation

### ☑️Reviewer Checklist

### 🎉 GIF

![giphy](https://user-images.githubusercontent.com/38032787/176711666-64616d61-bbd4-4821-b9bb-510d93259e00.gif)
